_posts/2023-04-03-technical-roadmap-opensearch-extensibility.markdown

```diff
@@ -0,0 +1,160 @@
+
+We will replace the two distribution flavors of OpenSearch/Dashboards (currently a -min distribution without security or plugins) with a set of distribution manifests tailored for such purposes as log analytics and search. Each distribution will be represented by a manifest that can be assembled by downloading the artifacts from the public catalog for packaging purposes. We want to enable single-click installation of a collection of extensions, provide recommended distributions of OpenSearch for tailored purposes, let vendors create their favorite flavor of OpenSearch/Dashboards distribution easily, and add the capability to create enterprise-tailored distributions.
+
+## Future
+
+Extensions will make a lot of new, big ideas possible! Here are some of our favorites.
+
+### Hot swap extensions 
+
+OpenSearch/Dashboards bootstrap plugins at start time, and various parts of the system assume that plugins do not change at runtime. Requiring a cluster restart for all extensions is a severely crippling limitation on the path of ecosystem adoption of any significant number of extensions, primarily because cluster restarts mean stopping inbound traffic. We ensure that any extension point can be loaded or unloaded at runtime by making settings in the OpenSearch core dynamic and adding tools to support loading and unloading extensions at runtime without restarting OpenSearch/Dashboards nodes or the entire cluster. This creates the ability to add, upgrade, or remove an extension without the need to restart OpenSearch/Dashboards or connect a remote extension to an existing OpenSearch/Dashboards cluster.
+
+### Extensible document parsing
+
+JSON is, by far, the most popular input format for OpenSearch. JSON is humanly readable, so it is pretty easy to test and use for development. Additionally, the OpenSearch ecosystem is built around JSON with most benchmarks written in JSON and ingest connectors supporting JSON. However, JSON is much slower and more space-consuming than most binary formats, thus swapping JSON for another type may yield significant performance gains. We would like to make OpenSearch input formats extensible so that it is easier to add and test more formats. This was proposed in [OpenSearch #4559](https://github.com/opensearch-project/OpenSearch/issues/4559).
+
+### Security isolation
+
+With extensions being designed to run on a separate virtual machine (VM), we can introduce multiple options for isolating extensions, such as containers and Java runtimes (e.g. Firecracker, GraalVM, and EBPF). We can also provide a new secure default runtime, and solve the problem of Java Security Manager (JSM) deprecation. We can further secure across new boundaries, ensuring all messages are encrypted in transit and all data is encrypted at rest.
```

Comment:
Last sentence: Either "We can further extend security across new boundaries" or "We can further secure [noun] across new boundaries".